app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,365 @@
+  public static final String ACTION_TURN_OFF_AFTER_O = "Turn_off_hotspot_after_oreo";
+  private final String IP_STATE_KEY = "ip_state_key";
+  private static final int MY_PERMISSIONS_ACCESS_FINE_LOCATION = 102;
+  private Intent serviceIntent;
+  private Task<LocationSettingsResponse> task;
+  boolean flag = false;
+  String ip;
+  String TAG = ZimHostActivity.this.getClass().getSimpleName();
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_zim_host);
+
+    setUpToolbar();
+
+    startServerButton = (Button) findViewById(R.id.startServerButton);
+    serverTextView = (TextView) findViewById(R.id.server_textView);
+
+    if (savedInstanceState != null) {
+      serverTextView.setText(
+          getString(R.string.server_started_message) + " " + savedInstanceState.getString(
+              IP_STATE_KEY));
+      startServerButton.setText(getString(R.string.stop_server_label));
+    }
+
+    FragmentManager fragmentManager = getSupportFragmentManager();
+    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
+    ZimFileSelectFragment fragment = new ZimFileSelectFragment();
+    fragmentTransaction.add(R.id.frameLayoutServer, fragment);
+    fragmentTransaction.commit();
+
+    serviceIntent = new Intent(this, HotspotService.class);
+    Context context = this;
+
+    startServerButton.setOnClickListener(new View.OnClickListener() {
+      @Override public void onClick(View v) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          toggleHotspot();
+        } else {
+          //TO DO: show Dialog() + within that add check mobile Data check later.
+          //if (isMobileDataEnabled(context)) {
+          //  mobileDataDialog();
+          //} else {
+          if (flag) {
+            serverStopped();
+            isStarted = false;
+          } else {
+            startHotspotDialog();
+          }
+          //}
+        }
+      }
+    });
+  }
+
+  @RequiresApi(api = Build.VERSION_CODES.O)
+  void toggleHotspot() {
+    //Check if location permissions are granted
+    if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
+        == PackageManager.PERMISSION_GRANTED) {
+      if (checkHotspotState(this)) //If hotspot is already enabled, turn it off
+      {
+        startService(ACTION_TURN_OFF_AFTER_O);
+      } else //If hotspot is not already enabled, then turn it on.
+      {
+        setupLocationServices();
+      }
+    } else {
+      //Ask location permission if not granted
+      ActivityCompat.requestPermissions(this,
+          new String[] { Manifest.permission.ACCESS_FINE_LOCATION },
+          MY_PERMISSIONS_ACCESS_FINE_LOCATION);
+    }
+  }
+
+  // This method checks if mobile data is enabled in user's device.
+  static boolean isMobileDataEnabled(Context context) {
+    boolean enabled = false;
```

Comment:
you don't need this variable, just return where you write/read it

Reply:
Fixed. 815ad025d77f89d5b2def01cf544cbb80bdfa5ca

Reply:
@iadeelzafar Can we do this without reflection? @macgills Should we keep it if using reflection is the only way?

Reply:
from my discussions with iadeel I think it is unfortunately necessary, if the reflection fails we pursue alternatives